File: api/schemas/factor.py
from pydantic import BaseModel
from typing import Optional, List


class FactorAnalyzeRequest(BaseModel):
    expression: str
    universe: str = "hs300"
    start_date: Optional[str] = None
    end_date: Optional[str] = None


class FactorAnalyzeResult(BaseModel):
    ic_mean: float
    ic_std: float
    icir: float
    rank_ic_mean: float
    turnover: float
    ic_series: List[float] = []
    returns: List[float] = []
    dates: List[int] = []

File: frontend/src/api/factor.ts
import { get, post } from './index'

export interface FactorAnalyzeRequest {
  factor_name: string
  expression: string
  universe?: string
  start_date?: string
  end_date?: string
}

export interface FactorAnalyzeResult {
  ic_mean: number
  ic_std: number
  icir: number
  rank_ic_mean: number
  turnover: number
  dates?: string[]
  ic_series?: number[]
  returns?: number[]
}

export const factorApi = {
  analyze: (data: FactorAnalyzeRequest) =>
    post<FactorAnalyzeResult>('/factor/analyze', data),

  getMetrics: (name: string) =>
    get<FactorAnalyzeResult>(`/factor/${name}/metrics`),
}
